manpages/strukt.1
.TH STRUKT "7" "June 2013" "CompatibleOne" "Platform"
.SH NAME
strukt \- CompatibleOne Structure Constructor
.SH SYNOPSIS
\fBstrukt\fR
.PP
.SH DESCRIPTION
.\" Add any additional description here
.PP
This tool allows to create structures to generate code for the CompatibleOne Accords plateform.
.SH OPTIONS
.TP
\fB\--verbose\fR
active verbose messages
.TP
\fB\--debug\fR
activate debug messages 
.TP
\fB\--echo\fR
echo source during input  
.TP
\fB\--xml\fR
generate xml production   
.TP
\fB\--occi\fR
generate occi production  
.TP
\fB\--mysql\fR
generate for mysql engine 
.TP
\fB\--crud\fR
generate crud actions     
.TP
\fB\--schema\fR
generate cords schema     
.TP
\fB\--cords <1:k|2:l|3:a|4:m>\fR
cords entity object type  
.SH AUTHOR
The CompatibleOne project.
.SH MANPAGE
Most of this documents comes from the README of ACCORDS, written by Iain James Marshall.
Written by Vincent Renardias <vrenardias@citypassenger.com> for the CompatibleOne project.
.SH "REPORTING BUGS"
Report CompatibleOne bugs to compatibleone@ow2.org
.br
CompatibleOne home page: <http://www.compatibleone.org>
.SH COPYRIGHT
Copyright \(co 2013 CompatibleOne.
License Apache 2.0.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


debian/changelog
compatibleone-accords (1.0-10) unstable; urgency=low

  * Add compatibleone-oaprocci package.
  * fix to build from the new github repository.
  * move comodel to base package and also include strukt.

 -- Vincent Renardias <vincent@renardias.com>  Thu, 06 Jun 2013 10:42:12 +0200

compatibleone-accords (1.0-9) unstable; urgency=low

  * Upstream code update (new codegen code).

 -- Vincent Renardias <vincent@renardias.com>  Mon, 06 May 2013 15:45:02 +0200

compatibleone-accords (1.0-8) unstable; urgency=low

  * Upstream code update (big dashboard scripts update).

 -- Vincent Renardias <vincent@renardias.com>  Mon, 06 May 2013 15:44:46 +0200

compatibleone-accords (1.0-7) unstable; urgency=low

  * Upstream code update.
  * New package: compatible-ec2procci.

 -- Vincent Renardias <vincent@renardias.com>  Mon, 18 Mar 2013 15:45:01 +0100

compatibleone-accords (1.0-6) unstable; urgency=low

  * Improved Python bindings.
  * Add 'compatibleone-cnprocci' package for ComputeNext interface.
  * Add 'compatibleone-eziprocci' package for EasyClouds interface.
  * Upstream code update.

 -- Vincent Renardias <vincent@renardias.com>  Mon, 11 Mar 2013 10:34:58 +0100

compatibleone-accords (1.0-5) unstable; urgency=low

  * Upstream update.
  * Fix dashboard scripts.

 -- Vincent <vincent@renardias.com>  Thu, 21 Feb 2013 22:17:00 +0100

compatibleone-accords (1.0-4) unstable; urgency=low

  * Add new package compatibleone-dashboard (more work needed).
  * Clean-up control file, start adding the php dashboard
  * Replace hardwired emails.

 -- Vincent <vincent@renardias.com>  Thu, 21 Feb 2013 18:49:00 +0100

compatibleone-accords (1.0-3) unstable; urgency=low

  * Upstream update (tree SHA: a5c42c620e43d7a698c35e725cb7cf7fbba61181)
  * Add libcocspi0 and libcocspi-dev packages (+bump up solib deps).

 -- Vincent <vincent@renardias.com>  Wed, 13 Feb 2013 12:18:57 +0100

compatibleone-accords (1.0-2) unstable; urgency=low

  * Upstream code of jan 23rd.
  * Include selftest.csp cordscript.
  * Remove occurences of paprocci in current build.

 -- Vincent <vincent@renardias.com>  Wed, 23 Jan 2013 14:58:01 +0100

compatibleone-accords (1.0-1) unstable; urgency=low

  * Bump up version number!
  * Update solib deps accordingly.
  * Include code for the current repository.
  * Code / Comments clean-ups in debian/rules.
  * Disable python testsuite (doesn't work and most likely won't be maintained).
  * Wrote minimalist manpage for the cosacs client to shut up lintian.
  * Declare compatibleone-accords and compatibleone-accords-dev as metapackages.
  * Packages are now fully clean according to lintian (except the Python JDK).

 -- Vincent Renardias <vincent@renardias.com>  Fri, 18 Jan 2013 14:01:40 +0100

compatibleone-accords (0.20130101-1) unstable; urgency=low

  * Upstream code update + .csp sample file.
  * Set dpkg-source 1.0 format.
  * Update lib soname dep.

 -- Vincent Renardias <vincent@renardias.com>  Wed, 02 Jan 2013 10:24:36 +0100

compatibleone-accords (0.20121218-1) unstable; urgency=low

  * New upstream code + more doc + sync all binary packages.

 -- Vincent Renardias <vincent@renardias.com>  Tue, 18 Dec 2012 14:43:00 +0200

compatibleone-accords (0.20121214-1) unstable; urgency=low

  * Several bug fixes regarding cosacs.

 -- Vincent Renardias <vincent@renardias.com>  Fri, 14 Dec 2012 16:58:00 +0200

compatibleone-accords (0.20121213-2) unstable; urgency=low

  * Add 'cosacs' package (client side COSACS to be run in the VM images).
  * Split 'compatibleone' in 'compatibleone-base' and 'compatibleone'.

 -- Vincent Renardias <vincent@renardias.com>  Thu, 13 Dec 2012 17:47:00 +0200

compatibleone-accords (0.20121213-1) unstable; urgency=low

  * Major renaming of packages to be fully Debian compliant and prepare the
    integration of the Cosacs client.
  * Force lib # in the dependency and bump the version to make sure we're
    not linking against a version older than today (lib ABI change).

 -- Vincent Renardias <vincent@renardias.com>  Tue, 12 Dec 2012 15:16:00 +0200

compatibleone-accords (0.20121212-1) unstable; urgency=low

  * Upstream update.
  * Major packaging update: each lib now has its own packages.

 -- Vincent Renardias <vincent@renardias.com>  Tue, 11 Dec 2012 15:47:00 +0200

compatibleone-accords (0.20121207-1) unstable; urgency=low

  * Upstream update.
  * Lots of packaging fixes (see gitorious log for the gory details)

 -- Vincent Renardias <vincent@renardias.com>  Fri, 07 Dec 2012 10:25:21 +0200

compatibleone-accords (0.20121129-1) unstable; urgency=low

  * Upstream update.

 -- Vincent Renardias <vincent@renardias.com>  Mon, 26 Nov 2012 11:43:34 +0200

compatibleone-accords (0.20121121-1) unstable; urgency=low


comodel/src/strukt.c
/* -------------------------------------------------------------------- */
/*  ACCORDS PLATFORM                                                    */
/*  (C) 2011 by Iain James Marshall (Prologue) <ijm667@hotmail.com>     */
/* -------------------------------------------------------------------- */
/* Licensed under the Apache License, Version 2.0 (the "License"); 	*/
/* you may not use this file except in compliance with the License. 	*/
/* You may obtain a copy of the License at 				*/
/*  									*/
/*  http://www.apache.org/licenses/LICENSE-2.0 				*/
/*  									*/
/* Unless required by applicable law or agreed to in writing, software 	*/
/* distributed under the License is distributed on an "AS IS" BASIS, 	*/
/* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or 	*/
/* implied. 								*/
/* See the License for the specific language governing permissions and 	*/
/* limitations under the License. 					*/
/* -------------------------------------------------------------------- */

#ifndef	_strukt_c
#define	_strukt_c

#include "strukt.h"

public	struct	strukt_context C = 	{ 
		0,		/* schema */
		(char *) 0,	/* license */
		(char *) 0,	/* scheme */
		(char *) 0,	/* klass */
		(char *) 0,	/* rel */
		"",		/* prefix */
		0,0,0,0,
		0,0,0,0,
		0,0,
		0,0,0,0,
		0,0,
		(FILE *) 0,
		(char *) 0,
		(struct item *) 0,
		(struct item *) 0
		};

#include "item.c"

public	void	reset_strukt()
{
	struct	item * iptr;
	while ((iptr=C.first) != (struct item *) 0)
	{
		C.first = iptr->next;
		liberate_item( iptr );
	}
	C.first = C.last = (struct item *) 0;
	if ( C.license ) C.license = liberate( C.license );
	if ( C.scheme ) C.scheme = liberate( C.scheme );
	if ( C.klass ) C.klass = liberate( C.klass );
	if ( C.rel ) C.rel = liberate( C.rel );
	memset( &C, 0, sizeof( struct strukt_context ));
	return;
}

public	int	strukt_set_license(char * filename)
{
	if (!( C.license = allocate_string(filename) ))
		return(27);
	else	return(0);
}

public	void	strukt_set_prefix(char * value )
{
	C.prefix = allocate_string( value );
}

public	void	strukt_set_gencrud(int value)
{
	C.gencrud = value;
	return;
}

public	void	strukt_set_genxml(int value)
{
	C.genxml = value;
	return;
}

public	void	strukt_set_schema(int value)
{
	C.schema = value;
	return;
}

public	void	strukt_set_genrest(int value)
{
	C.genrest = value;
	return;
}

public	void	strukt_set_genocci(int value)
{
	C.genocci = value;
	return;
}

private	void	expand( FILE * h, char * sptr )
{
	int	c;
	while ((c = *(sptr++)) != 0)
		fprintf(h,"%c ",c);
	return;
}

private	void	gener( FILE * h, int c ,int n )
{
	while ( n-- ) fprintf(h,"%c",c);		
	return;
}

public	void	title( FILE * h, char * sptr )
{
	fprintf(h,"\n/*\t");
	gener( h, '-', strlen( sptr ) * 2 );
	fprintf(h,"\t*/\n");

	fprintf(h,"/*\t");
	expand( h, sptr );
	fprintf(h,"\t*/\n");

	fprintf(h,"/*\t");
	gener( h, '-', strlen( sptr ) * 2 );
	fprintf(h,"\t*/\n");
	return;
}

private	void	generate_reset( FILE *h, char * nptr )
{
	int	i;
	struct	item * iptr;
	char	buffer[512];
	sprintf(buffer,"%s_%s",nptr,C.name);
	title(h, buffer );
	fprintf(h,"public struct %s * %s(struct %s * sptr)\n{\n",C.name,buffer,C.name);
	fprintf(h,"\tif ( sptr )\n\t{\n");
	for ( 	iptr= C.first;
		iptr != (struct item *) 0;
		iptr = iptr->next )
	{
		if (!( iptr->dimension ))
		{
			fprintf(h,"\t\tsptr->%s = ",iptr->name);
			if ( iptr->indirection )
			{
				fprintf(h,"(");
				if (!( strcmp( iptr->basic,"struct" ) ))
					fprintf(h,"struct %s",iptr->type);
				else	fprintf(h,"%s",iptr->basic);
				for ( i=0;i < iptr->indirection; i++ )
						fprintf(h,"*");
				fprintf(h,") 0;\n");
			}
			else	fprintf(h," 0;\n");
		}
		else	
		{
			fprintf(h,"\t\tmemset(&sptr->%s,0,sizeof(",iptr->name);
			if (!( strcmp( iptr->basic,"struct" ) ))
				fprintf(h,"struct %s",iptr->type);
			else	fprintf(h,"%s",iptr->basic);
			for ( i=0;i < iptr->indirection; i++ )
				fprintf(h,"*");
			fprintf(h,")*%u);\n",iptr->dimension);
		}
	}
	fprintf(h,"\t}\n");
	fprintf(h,"\treturn(sptr);\n");
	fprintf(h,"\n}\n");

}

private	void	generate_allocate( FILE *h, char * nptr )
{
	int	i;
	struct	item * iptr;
	char	buffer[512];
	sprintf(buffer,"%s_%s",nptr,C.name);
	title(h, buffer );
	fprintf(h,"public struct %s * %s()\n{\n",C.name,buffer);
	fprintf(h,"\tstruct %s * sptr;\n",C.name);
	fprintf(h,"\tif (!( sptr = allocate( sizeof( struct %s ) ) ))\n\t\treturn( sptr );\n",C.name);
	fprintf(h,"\telse\treturn( reset_%s(sptr) );",C.name);
	fprintf(h,"\n}\n");
}

private	void	generate_add( FILE *h, char * nptr )
{
	int	i;
	struct	item * iptr;
	char	buffer[512];
	sprintf(buffer,"%s_%s",nptr,C.name);
	title(h, buffer );
	fprintf(h,"public struct %s * %s(struct %s * pptr)\n{\n",C.name,buffer,C.parent);
	fprintf(h,"\tstruct %s * sptr;\n",C.name);
	fprintf(h,"\tif (!( sptr = allocate( sizeof( struct %s ) ) ))\n\t\treturn( sptr );\n",C.name);
	fprintf(h,"\telse if (!( sptr = reset_%s(sptr) ))\n\t\treturn( sptr );\n",C.name);
	fprintf(h,"\telse\n\t{\n");
	fprintf(h,"\t\tif (!( sptr->previous = pptr->last ))\n");
	fprintf(h,"\t\t\tpptr->first = sptr;\n");
	fprintf(h,"\t\telse\tsptr->previous->next = sptr;\n");
	fprintf(h,"\t\tpptr->last = sptr;\n");
	fprintf(h,"\t\tsptr->parent = pptr;\n");
	fprintf(h,"\t\treturn( sptr );\n");
	fprintf(h,"\t}\n");
	fprintf(h,"\n}\n");
}

private	void	generate_drop( FILE *h, char * nptr )
{
	int	i;
	struct	item * iptr;
	char	buffer[512];
	sprintf(buffer,"%s_%s",nptr,C.name);
	title(h, buffer );
	fprintf(h,"public struct %s * %s(struct %s * sptr)\n{\n",C.name,buffer,C.name);
	fprintf(h,"\tif ( sptr )\n\t{\n");
	fprintf(h,"\t\tif (!( sptr->parent )) return(sptr);\n");

		fprintf(h,"\t\tif (!( sptr->previous ))\n");
		fprintf(h,"\t\t{\n");
			fprintf(h,"\t\t\tif (!( sptr->parent->first = sptr->next ))\n");
				fprintf(h,"\t\t\t\tsptr->parent->last = (struct %s *) 0;\n",C.name);
			fprintf(h,"\t\t\telse	sptr->parent->first->previous = (struct %s *) 0;\n",C.name);
		fprintf(h,"\t\t}\n");
		fprintf(h,"\t\telse if (!( sptr->previous->next = sptr->next ))\n");
			fprintf(h,"\t\t\tsptr->parent->last = sptr->previous;\n");

		fprintf(h,"\t\tif (!( sptr->next ))\n"); 
		fprintf(h,"\t\t{\n");
			fprintf(h,"\t\t\tif (!( sptr->parent->last = sptr->previous ))\n");
				fprintf(h,"\t\t\t\tsptr->parent->first = (struct %s *) 0;\n",C.name);
			fprintf(h,"\t\t\telse	sptr->parent->last->next = (struct %s *) 0;\n",C.name);
		fprintf(h,"\t\t}\n");
		fprintf(h,"\t\telse if (!( sptr->next->previous = sptr->previous ))\n");
			fprintf(h,"\t\t\tsptr->parent->first = sptr->next;\n");
		fprintf(h,"\t\tsptr = liberate_%s(sptr);\n",C.name);

	fprintf(h,"\t}\n");
	fprintf(h,"\treturn((struct %s *) 0);",C.name);
	fprintf(h,"\n}\n");
}

private	void	generate_occi_request2file( FILE * h, char * nptr )
{
	char	* prefix;
	int	item=0;
	char	buffer[512];
	struct	item * iptr;
	prefix = "c1";
	sprintf(buffer,"%s_%s",nptr,C.name);
	title(h, buffer );
	fprintf(h,"public int %s(FILE * fh,struct %s * sptr,char * prefix, char * nptr)\n{\n",buffer,C.name);
	fprintf(h,"\tstruct xml_element * wptr;\n");
	fprintf(h,"\tif (!( sptr )) return(0);\n");
	fprintf(h,"\tfprintf(fh,%cPOST /%cs/ HTTP/1.1\\r\\n%c,nptr);\n",0x0022,0x0025,0X0022);
	fprintf(h,"\tfprintf(fh,%cCategory: %cs; scheme='http://scheme.%cs.org/occi/%cs#'; class='kind';\\r\\n%c,nptr,prefix,prefix);\n",
		0x0022, 0x0025, 0x0025,0x0025,0x0022);
	for ( 	iptr= C.first;
		iptr != (struct item *) 0;
		iptr = iptr->next )
	{
		if (!( strcmp( iptr->basic,"struct" ) ))
			continue;
		else if ( iptr->indirection )
		{
			fprintf(h,"\tfprintf(fh,%cX-OCCI-Attribute: %cs.%cs.%s='%cs'\\r\\n%c,prefix,nptr,(sptr->%s?sptr->%s:%c%c));\n",
				0x0022,0x0025,0x0025,iptr->name,0x0025,0x0022,iptr->name,iptr->name,0x0022,0x0022);
		}
		else 
		{
			fprintf(h,"\tfprintf(fh,%cX-OCCI-Attribute: %cs.%cs.%s='%cu'\\r\\n%c,prefix,nptr,sptr->%s);\n",
				0x0022,0x0025,0x0025,iptr->name,0x0025,0x0022,iptr->name);
		}
	}
	fprintf(h,"\treturn(0);\n");
	fprintf(h,"\n}\n");
}

private	void	generate_interpret_xml( FILE * h, char * nptr )
{
	char	* prefix;
	int	item=0;
	char	buffer[512];
	struct	item * iptr;
	sprintf(buffer,"%s_%s",nptr,C.name);
	title(h, buffer );
	fprintf(h,"public int %s(struct %s * sptr,struct xml_element * eptr)\n{\n",buffer,C.name);
	fprintf(h,"\tstruct xml_element * wptr;\n");
	fprintf(h,"\tif (!( eptr )) return(0);\n");
	fprintf(h,"\tif (!( sptr )) return(0);\n");
	fprintf(h,"\tfor ( wptr=eptr->first; wptr != (struct xml_element *) 0; wptr=wptr->next)\n\t{\n");
	prefix = "\0";
	for ( 	iptr= C.first;
		iptr != (struct item *) 0;
		iptr = iptr->next )
	{
		if (!( strcmp( iptr->basic,"struct" ) ))
			continue;
		fprintf(h,"\t\t%sif (!( strcmp(wptr->name,%c%s%c) ))\n",prefix,0x0022,iptr->name,0x0022);
		prefix = "else ";
		fprintf(h,"\t\t{\n");
		if ( iptr->indirection )
		{
			fprintf(h,"\t\t\tif ( wptr->value ) { sptr->%s = allocate_string(wptr->value); }\n",iptr->name);
		}
		else
		{
			fprintf(h,"\t\t\tif ( wptr->value ) { sptr->%s = atoi(wptr->value); }\n",iptr->name);
		}
		fprintf(h,"\t\t}\n");
	}
	fprintf(h,"\t}\n");
	fprintf(h,"\treturn(0);\n");
	fprintf(h,"\n}\n");
}




private	void	generate_liberate( FILE *h, char * nptr )
{
	char	buffer[512];
	struct	item * iptr;
	sprintf(buffer,"%s_%s",nptr,C.name);
	title(h, buffer );
	fprintf(h,"public struct %s * %s(struct %s * sptr)\n{\n",C.name,buffer,C.name);
	if ( C.hasfirst ) 
	{
		fprintf(h,"\tstruct %s * mptr;\n",C.premier);
	}
	fprintf(h,"\tif ( sptr )\n\t{\n");
	for ( 	iptr= C.first;
		iptr != (struct item *) 0;
		iptr = iptr->next )
	{
		if ( C.hasfirst ) 
		{
			if (!( strcmp( iptr->name, "last" ) ))
				continue;
			if (!( strcmp( iptr->name, "first" ) ))
			{
				fprintf(h,"\t\twhile ((mptr = sptr->first) != (struct %s *) 0)\n\t\t{\n",C.premier);
				fprintf(h,"\t\t\tsptr->first = mptr->next;\n");
				fprintf(h,"\t\t\tmptr = liberate_%s( mptr );\n\t\t}\n",C.premier);
				continue;
			}
		}
		if (!( strcmp( iptr->name, "parent" ) ))
			continue;
		else if (!( strcmp( iptr->basic, "void" ) ))
			continue;
		else if ( iptr->indirection == 1 )
		{
			if (!( strcmp( iptr->name, "previous" )))
				continue;
			else if (!( strcmp( iptr->name, "next" )))
				continue;
			fprintf(h,"\t\tif ( sptr->%s )\n\t\t\t sptr->%s = liberate",iptr->name,iptr->name);
			if (!( strcmp( iptr->basic, "struct" ) ))
				fprintf(h,"_%s",iptr->type);
			fprintf(h,"(sptr->%s);\n",iptr->name);
		}
	}
	fprintf(h,"\t\tsptr = liberate( sptr );\n");
	fprintf(h,"\t}\n");
	fprintf(h,"\treturn((struct %s *) 0);\n",C.name);
	fprintf(h,"\n}\n");
}

public	int	generate()
{
	struct	item * iptr;
	if ( C.schema )
	{
		fprintf(C.target,"\n<xsd:complexType name=%c%s%c>\n",0x0022,C.name,0x0022);
		for ( 	iptr= C.first;
			iptr != (struct item *) 0;
			iptr = iptr->next )
		{
			fprintf(C.target,"\t<xsd:attribute name=%c%s%c",0x0022,iptr->name,0x0022);
			if (!( strcmp( iptr->basic , "int" ) ))
				fprintf(C.target," type=%c%s%c",0x0022,"xsd:integer");
			else	fprintf(C.target," type=%c%s%c",0x0022,"xsd:string");
			fprintf(C.target,"/>\n");
		}
		fprintf(C.target,"</xsd:complexType>\n");
	}
	else if ( C.genrest )
	{
		switch ( C.genrest )
		{
		case	_OCCI_KIND	:
			C.klass  = allocate_string("kind");
			C.scheme = allocate_string("http://scheme.compatibleone.fr/scheme/compatible#");
			C.rel    = allocate_string("http://scheme.ogf.org/occi/resource#");
			break;
		case	_OCCI_LINK	:
			C.klass  = allocate_string("link");
			C.scheme = allocate_string("http://scheme.compatibleone.fr/scheme/compatible#");
			C.rel    = allocate_string("http://scheme.ogf.org/occi/link#");
			break;
		case	_STRUKT_OCCI_ACTION	:
			C.klass  = allocate_string("action");
			C.scheme = allocate_string("http://scheme.compatibleone.fr/scheme/compatible#");
			C.rel    = allocate_string("http://scheme.ogf.org/occi/mixin#");
			break;
		case	_STRUKT_OCCI_MIXIN	:
			C.klass  = allocate_string("mixin");
			C.scheme = allocate_string("http://scheme.compatibleone.fr/scheme/compatible#");
			C.rel    = allocate_string("http://scheme.ogf.org/occi/action#");
			break;
		default	:
			return( failure(30,"incorrect","cords value") );
		}
		if ( C.gensql )
		{
			generate_occi_sql_builder( C.target, "occi" );
		}
		else
		{
			generate_occi_rest_builder( C.target, "occi" );
			generate_occi_rest_request( C.target, "rest_occi" );
		}
	}
	else
	{
		generate_liberate( C.target, "liberate" );
		generate_reset( C.target, "reset" );
		generate_allocate( C.target, "allocate" );
		if (( C.hasparent ) 
		&&  ( C.hasprevious )
		&&  ( C.hasnext ))
		{
			generate_add( C.target, "add" );
			generate_drop( C.target, "drop" );
		}
		if ( C.genxml )
		{
			generate_interpret_xml( C.target, "xmlin" );
		}
		if ( C.genocci )
		{
			generate_occi_request2file( C.target, "rest_occi" );
		}
	}
	return(0);
}

private	int	is_punctuation(int c)
{
	if (( c >= 0x0021 ) && ( c <= 0x002F ))
		return(1);
	else if (( c >= 0x003a ) && ( c <= 0x003f ))
		return(1);
	else if (( c >= 0x005b) && ( c <= 0x005e ))
		return(1);
	else if (( c >= 0x007b) && ( c <= 0x007f ))
		return(1);
	else	return(0);
}

private	int	ungetch(int c)
{
	return((C.ungotc = c));
}

private	int	getch(FILE *h)
{
	int	c;
	if ((c = C.ungotc) != 0)
		C.ungotc = 0;
	else if ((c = fgetc(h)) == -1)
		c = 0;
	else if ( C.echo )
		printf("%c",c);
	return( c );
}

private	int	is_white( int c )
{
	switch ( c )
	{
	case	' '	:
	case	'\r'	:
	case	'\t'	:
	case	'\n'	: return( c );
	default		: return( 0 );
	}
}

private	int	remove_white_space(FILE * h)
{
	int	c;
	while (1)
	{
		switch ((c = getch(h)))
		{
		case	0x23	:
			while (((c = getch(h)) != 0) && ( c != '\n'));
			if ( c  == '\n' )
				continue;
		case	0	:
		case	26	: 
			return(0);
		default	:
			if ( is_white( c ) )
				continue;
			else	return(ungetch(c));
		}
	}

}

private	int	get_token( FILE * h, char * token, int tlen )
{
	int	c;
	int	i=0;
	int	quoting=0;
	while ( i < tlen )
	{
		if (!(c = getch(h)))
			break;
		else if ( c == '"' )
		{
			quoting = c;
			continue;
		}
		else if ( c == quoting ) 
		{
			quoting = 0;
			break;
		}
		else if ( quoting )
		{
			*(token+i) = c;
			i++;
			continue;
		}
		else if ( is_white( c ) )
		{
			ungetch( c );
			break;
		}
		else if ( is_punctuation( c ) )
		{
			ungetch( c );
			break;
		}
		else
		{
			*(token+i) = c;
			i++;
			continue;
		}
	}
	*(token+i) = 0;
	return(i);
}

public	int	add_member()
{
	struct	item * iptr;
	if (!( iptr = allocate_item()))
		return( failure( 27,"allocation","item") );
	else if (!( iptr->name = allocate_string( C.member )))
		return( failure( 27,"allocation","item->name") );
	else if (!( iptr->basic = allocate_string( C.basic )))
		return( failure( 27,"allocation","item->basic") );
	else 	
	{
		if (!( strcmp( C.basic, "struct" ) ))
			if (!( iptr->type = allocate_string( C.special )))
				return( failure( 27,"allocation","item->type") );
		iptr->dimension = C.dimension;
		iptr->indirection = C.isptr;
		C.isptr = 0;
		C.dimension=0;
		if (!( iptr->previous = C.last ))
			C.first = iptr;
		else	iptr->previous->next = iptr;
		C.last = iptr;
		if (!( strcmp( C.basic, "struct" ) )) 
		{
			if (!( strcmp(C.member, "first" ) ))
			{
				C.hasfirst=1;
				strcpy( C.premier, C.special );
			}
			else if (!( strcmp(C.member, "parent" ) ))
			{
				C.hasparent=1;
				strcpy( C.parent, C.special );
			}
			else if (!( strcmp(C.member, "previous" ) ))
			{
				C.hasprevious=1;
			}
			else if (!( strcmp(C.member, "next" ) ))
			{
				C.hasnext=1;
			}
		}
		return( 0 );
	}
}

private	int	handle_comment( FILE * sh )
{
	int	c;
	int	state=0;
	switch ( getch(sh) )
	{
	case	'/'	:
		while (((c = getch(sh)) > 0) && ( c != '\n'));
		break;
	case	'*'	:
		while ((c = getch(sh)) > 0)
		{
			if (!( state ))
			{
				if ( c == '*' )
					state = 1;
			}
			else if ( c == '/' )
				break;
			else	state = 0;
		}
		break;

	}
	return(0);
}	

private	int	handle_punctuation( FILE * sh, int c )
{
	int	status;
	char 	b[2];
	switch ( C.state )
	{
	case	0 :	/* awaiting struct */
		if ( c == '/' )
		{
			handle_comment(sh);
			return(0);
		}
			
	case	1 :	/* awaiting struct name */
		break;
	case	2 :
		if ( c == '{' )
		{
			C.state = C.next;
			return( 0 );
		}
		else	break;

	case	3 :	/* awaiting type keyword */
		if ( c == '}' )
		{
			C.state = 9;
			return(0);
		}
		else if ( c == '/' )
		{
			handle_comment(sh);
			return(0);
		}
		else	break;

	case	4 :	/* awaiting extended type */
	case	5 :	/* awaiting name or pointer */
		if ( c == '*' )
		{
			C.isptr++;
			return(0);
		}
		break;
	case	6 :	/* awaiting colon or dimension */
		if ( c == ';' )
		{
			if (!( status = add_member() ))
			{
				C.state = C.next;
				return( 0 );
			}
			else	break;
		}
		else if ( c == '[' )
		{
			C.state = 7;
			return( 0 );
		}
		else	break;
	case	8 :	/* awaiting closure of dimension */
		if ( c == ']' )
		{
			C.state = 6;
			return(0);
		}
		else	break;
	case	9 :	/* awaiting colon or dimension */
		if ( c == ';' )
		{
			C.state = 0;
			return(generate());
		}
		else if ( c == '/' )
		{
			handle_comment(sh);
			return(0);
		}
	}
	b[0] = c;
	b[1] = 0;
	return(failure(30,"incorrect punctuation",b));
}

private	char *	is_basic_type(char * tptr)
{
	if (!(strcmp( tptr,"struct")))
		return( "struct" );
	else if (!(strcmp( tptr,"float")))
		return( "float" );
	else if (!(strcmp( tptr,"double")))
		return( "double" );
	else if (!(strcmp( tptr,"char")))
		return( "char" );
	else if (!(strcmp( tptr,"int")))
		return( "int" );
	else if (!(strcmp( tptr,"short")))
		return( "short" );
	else if (!(strcmp( tptr,"long")))
		return( "long" );
	else if (!(strcmp( tptr,"void")))
		return( "void" );
	else	return( (char *) 0);
}

private	int	handle_token( FILE * sh, char * tptr )
{
	switch ( C.state )
	{
	case	0 :	/* awaiting struct */
		if (!( strcmp( tptr, "struct" ) ))
		{
			C.state = 1;
			return(0);
		}
		else	break;
	case	1 :	/* awaiting struct name */
		strcpy( C.name, tptr );
		C.next = 3;
		C.state = 2;
		return(0);
	case	2 :	/* awaiting opener */
		break;
	case	3 :	/* awaiting type */
		if ((C.basic = is_basic_type( tptr )) != (char *) 0)
		{
			if (!(strcmp( C.basic,"struct")))
				C.state = 4;
			else	C.state = 5;
			return(0);
		}
		else	break;

	case 	4 :	/* awaiting type name */
		strcpy(C.special,tptr);
		C.state = 5;
		return(0);
	case 	5 :	/* awaiting name */
		strcpy(C.member,tptr);
		C.state = 6;
		return(0);
	case	6 :	/* awaiting semi or dimensions */
		break;
	case	7 :	/* awaiting dimension */
		C.dimension = atoi( tptr );
		C.state = 8;
		return(0);
	}
	return( failure( 30 , "incorrect token", tptr ) );
}


private	void	file_symbol( FILE * h, char * sptr )
{
	int	c;
	fprintf(h,"_");
	while ((c = *(sptr++)) != 0)
	{
		if ( is_punctuation( c ) )
			fprintf(h,"_");
		else	fprintf(h,"%c",c);
	}
	fprintf(h,"_");
	return;
}

private	void	prepend_license( FILE * target, char * filename )
{
	int	c;
	FILE * h;
	if ((h = fopen( filename, "r")) != (FILE *) 0)
	{
		while ((c = fgetc(h)) > 0)
			fputc(c,target);
		fclose(h);
	}
	return;
}

private	void	schema_header( FILE * h, char * nptr, char * iptr)
{
	fprintf(h,"\n<xsd:sxhema xmlns:xsd=%chttp://www.w3c.org/2001/XMLSchema%c>\n",0x0022,0x0022);
	return;
}

private	void	schema_footer( FILE * h, char * nptr )
{
	fprintf(h,"\n</xsd:sxhema>\n");
	return;
}

private	void	file_header( FILE * h, char * nptr, char * iptr)
{
	char	buffer[1024];
	if ( C.genrest )
		sprintf(buffer,"occi%s",nptr);
	else	strcpy(buffer,nptr);
	if ( C.license ) { prepend_license( h, C.license ); }
	fprintf(h,"/* STRUKT WARNING : this file has been generated and should not be modified by hand */\n");
	fprintf(h,"#ifndef ");
	file_symbol( h, buffer );
	fprintf(h,"\n");
	fprintf(h,"#define ");
	file_symbol( h, buffer );
	fprintf(h,"\n");
	if ( C.genxml )
	{
		fprintf(h,"\n#include %c%s%c\n",0x0022,"element.h",0x0022);
	}
	fprintf(h,"\n#include %c%s%c\n",0x0022,iptr,0x0022);
	return;
}

private	void	file_footer( FILE * h, char * nptr)
{
	char	buffer[1024];
	if ( C.genrest )
		sprintf(buffer,"occi%s",nptr);
	else	strcpy(buffer,nptr);
	fprintf(h,"\n#endif\t/* ");
	file_symbol( h, buffer );
	fprintf(h," */\n");
	return;
}

public	int	process( char * nptr )
{
	int	status;
	int	c;
	int	l;
	char	sn[512];
	char	tn[512];
	char 	token[512];
	FILE * sh;
	sprintf(sn,"%s.h",nptr);
	sprintf(tn,"%s%s.c",C.prefix,nptr);
	if (!( sh = fopen( sn, "r" )))
		return( failure(40,"file not found",sn) );
	else if (!( C.target = fopen( tn, "w" )))
	{
		fclose(sh);
		return( failure(46,"creating file",tn) );
	}
	else
	{
		file_header( C.target, tn, sn );
		while ((c = remove_white_space( sh )))
		{
			if ( is_punctuation(c) )
			{
				if (!(status = handle_punctuation(sh,getch(sh))))
					continue;
				else	break;
			}
			else if (!( l = get_token(sh,token,512) ))
				break;
			else if (!( status = handle_token(sh,token) ))
				continue;
			else	break;
		}

		file_footer( C.target, tn );
		fclose(C.target);
		C.target = (FILE *) 0;
		fclose(sh);
	}		
	return( 0 );
}

public	int	schema( char * nptr )
{
	int	status;
	int	c;
	int	l;
	char	sn[512];
	char	tn[512];
	char 	token[512];
	FILE * sh;
	sprintf(sn,"%s.h",nptr);
	sprintf(tn,"%s.xsd",nptr);
	if (!( sh = fopen( sn, "r" )))
		return( failure(40,"file not found",sn) );
	else if (!( C.target = fopen( tn, "w" )))
	{
		fclose(sh);
		return( failure(46,"creating file",tn) );
	}
	else
	{
		schema_header( C.target, tn, sn );
		while ((c = remove_white_space( sh )))
		{
			if ( is_punctuation(c) )
			{
				if (!(status = handle_punctuation(sh,getch(sh))))
					continue;
				else	break;
			}
			else if (!( l = get_token(sh,token,512) ))
				break;
			else if (!( status = handle_token(sh,token) ))
				continue;
			else	break;
		}

		schema_footer( C.target, tn );
		fclose(C.target);
		fclose(sh);
	}		
	return( 0 );
}


	/* --------- */
#endif	/* _strukt_c */
	/* --------- */

